Backport support for mouse buttons 4 + 5 (#1076, #1068)

CendioHalim opened the pull request:
This is a backport of #1081 to branch-1.3.

Tested on X11, Wayland, macOS 15.0, Windows 10 & 11.

I'll leave this as a draft until #1081 is merged.

Albrecht-S commented:
@CendioHalim: PR #1081 has been merged in `master`. Finally there were two commits (the 2nd commit added by @MatthiasWM:
- commit 4f4a9be15b0c52837dd4d4a04ff021cc3e5d691e: "Support mouse buttons 4 + 5 ..."
- commit 6e3b64425e2b6345a4ba1d14c3c88f85bfc7f1a0: "Add fwd and back mouse buttons to test/keyboard"

You may want to update your PR. Feel free to rebase/squash your commit(s) so we have only one commit for branch-1.3. Thanks in advance.

CendioHalim commented:
Alright, I've added 6e3b64425e2b6345a4ba1d14c3c88f85bfc7f1a and rebased my branch now. Tested running the `keyboard` test and things look as expected.

CendioOssman commented on FL/Fl.H:
```diff
@@ -689,24 +691,35 @@ int main() {
+    | Mouse    | FL_BUTTON2     | middle button           |       |
+    | Mouse    | FL_BUTTON3     | right button            |       |
+    | Mouse    | FL_BUTTON4     | side button 1 (back)    | 1.4.0 |
+    | Mouse    | FL_BUTTON5     | side button 2 (forward) | 1.4.0 |
```
1.3.9?

Albrecht-S replied:
The original (1.4.0) docs list this as "since 1.4.0" because 1.3.9 did not yet include it. Once it will be backported to 1.3.x this should read "`1.3.10`".

I'll work on this PR after 1.4.0 RC3 or the final release, whichever comes first. I'd like to let the 1.4 code settle so we can see if there are any issues. The next 1.3 release should definitely be the last - unless unforeseen severe issues arise.

CendioHalim replied:
Sounds good, I'll update the table to say 1.3.10.

> I'd like to let the 1.4 code settle so we can see if there are any issues. 

Would you be able to say if the API will be stable, even if issues arise?

Albrecht-S commented:
> Would you be able to say if the API will be stable, even if issues arise?

Of course not! ;-)

It would really be a pain if the API needed to be changed. I'm pretty confident that it is OK, but you never know.

FWIW: as long as 1.4.0 is not finally released we are able to change the API but after the final release we're technically bound to the existing API, i.e. we can extend the functionality as usual, add parameters, etc. but not change it fundamentally. ... Unless we really need this to fix bugs that can't be fixed otherwise.

Albrecht-S commented:
@CendioHalim I'm meanwhile working on this, please don't update this PR anymore...

I'm going to fix some minor issues with this backport, and I'll probably push a commit to the PR before I can merge it.

Albrecht-S commented:
Mostly comments and documentation, but also a missing declaration we fixed for MinGW builds (475bfa92308b52d6fc307f376d617697f8d95ae2).

I also updated the widget order in `test/keyboard_ui.fl` (18bda5315d727f67ec55e47579e3625e78389d85).

Off the top of my head only the MinGW issue was essential.
I'll let you know the details when I can find more time to come back to this issue.

CendioHalim commented:
Ah, nice catch :)

Albrecht-S commented:
@CendioHalim I updated (and rebased) your branch and finally merged it into `branch-1.3`.

Please test and confirm that the tip of `branch-1.3` works for you. I intend to release 1.3.10 soon (probably after your confirmation). Thanks in advance and thanks for doing most of the backporting stuff.

Albrecht-S commented:
Looking forward to your test results.

CendioHalim commented:
Tested with TigerVNC on:
* x86 Linux
* armhf rpi5
* 32-bit Windows 10
* 64-bit Windows 10
* M1 mac osx 15.1

Works well for me, cheers :smile:

Albrecht-S commented:
@CendioHalim Awesome!

Thanks for the feedback and for testing on so many platforms!

Now I can go ahead and publish release 1.3.10 after some final updates.

Albrecht-S commented:
Done. [FLTK 1.3.10 has been released](https://www.fltk.org/articles.php?L1954).
